mkdocs_git_authors_plugin/repo.py

```diff
@@ -0,0 +1,774 @@
+                and len(self.authors()) > 1
+                else ''
+            )
+            authors_summary.append(
+                "<a href='mailto:%s'>%s</a>%s" % (
+                    author.email(),
+                    author.name(),
+                    contrib
+                ))
+        authors_summary = ', '.join(authors_summary)
+        return "<span class='git-authors'>%s</span>" % authors_summary
+
+    def _execute(self):
+        """
+        Execute git blame and parse the results.
+        """
+
+        cmd = GitCommand('blame', ['-lts', str(self._path)])
+        cmd.run()
+
+        # Retrieve SHA and content from the line, discarding
+        # file path and line number
+        line_pattern = re.compile('(.*?)\s.*\s*\d\)(\s*.*)')
+
+        for line in cmd.stdout():
+            m = line_pattern.match(line)
+            if m:
+                sha = m.group(1)
+                content = m.group(2).strip()
+
+                if content or self.repo().config('count_empty_lines'):
+                    # assign the line to a commit and count it
+                    commit = self.repo().commit(sha)
+                    author = commit.author()
+                    if author not in self._authors:
+                        self._authors.append(author)
+                    author.add_lines(self, commit)
+                    self.add_total_lines()
+                    self.repo().add_total_lines()
+
+    def path(self):
+        """
+        The path to the markdown file.
+
+        Args:
+
+        Returns:
+            Absolute path as Path object.
+        """
+        return self._path
+
+    def total_lines(self):
+        """
+        Total number of lines in the markdown source file.
+
+        Args:
+
+        Returns:
+            int
+        """
+        return self._total_lines
+
+
+class Author(AbstractRepoObject):
+# Sorted after Page for the function annotations
+    """
+    Abstraction of an author in the Git repository.
+    """
+
+    def __init__(self, repo: Repo, name: str, email: str):
+        """
+        Instantiate an Author.
+
+        Args:
+            repo: reference to the global Repo instance
+            name: author's full name
+            email: author's email
+        """
+        super().__init__(repo)
+        self._name = name
+        self._email = email
+        self._pages = {}
+
+    def add_lines(self, page: Page, commit: Commit, lines: int = 1):
+        """
+        Add line(s) in a given page/commit to the author's data.
+
+        Args:
+            page: Page object referencing the markdown file
+            commit: commit in which the line was edited (=> timestamp)
+            lines: number of lines to add. Default: 1
+        """
+        path = page.path()
+        entry = self.page(path, page)
+        entry['lines'] += lines
+        current_dt = entry.get('datetime')
+        commit_dt = commit.datetime()
+        if not current_dt or commit_dt > current_dt:
```

Comment:
I think I have made things more confusing than necessary by naming two different things the same: `page`.

Usually you'd expect `self.page()` to return a `Page` object, either existing or newly created.
`Author.page()` is different and should therefore be renamed (although I'm not quite sure to what).

The `Author` object stores information about how many lines it changed on which pages. This is stored in a dictionary with the page paths as key. Each entry holds a Page object, the line count, and the datetime of the last modification as a datetime object and as a string representation. But initially no information on the datetime is stored in the dictionary.

This function is called while processing the Git blame information. An author "receives" a line from a page in a given commit.
First we look up the entry maintaining the page. 

When the Commit object is created there are no lines yet, therefore `current_dt` is initialized with `None`. This logic ensures that in the first iteration the datetime is set to the commit's timestamp while in each subsequent execution of the function the commit's timestamp is compared to the entry's.

BUT (and I've been wondering about this all along): If I'm not mistaken this information is not really being used anywhere. It ends up in the `author` dict in the page context. But I can't really figure out who would make use of the information when *a given author* last modified the current page.
If we could get rid of that information in the `author` dict this could be written much more concisely, there wouldn't even be the need for the `commit` argument.

Reply:
>But I can't really figure out who would make use of the information when a given author last modified the current page. If we could get rid of that information in the author dict this could be written much more concisely

I'm still thinking about how to merge this plugin with `mkdocs-revision-date-localized`. I researched if retrieving datetime information from authors in `git blame` would solve the problem where shallow clones (used in CI builds, see https://github.com/timvink/mkdocs-git-revision-date-localized-plugin/issues/10) make it hard to find last update date of a file. [It doesn't](http://git.661346.n2.nabble.com/Shallow-clones-vs-blame-log-td6082351.html). 

I don't see any other use case. 

So let's decide not to supply the page authors last modification date, and make things a bit simpler!

Reply:
> I don't see any other use case.

That had been my impression, actually.

But I suggest *not* to decide this right now.
*If* we decide to keep the Git classes (vs. GitPython) *and* to merge the two plugins at some point we may well retrieve the information in this function. While I see that it doesn't solve the issue of shallow copies (and upon first sight I have the impression there is no remedy for that problem) it could make yet another invocation of Git obsolete.

Reply:
I agree, let's leave it for now. We can always revisit / refactor with future new insights.